lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp

```diff
@@ -1578,6 +1639,459 @@ class ConvertAtenScaledDotProductAttentionOp
+      extractedIdxMapExprs.push_back(rewriter.getAffineDimExpr(size.index()));
+    }
+
+    auto extractedIdxMaps = AffineMap::inferFromExprList(
+        {extractedIdxMapExprs, extractedIdxMapExprs}, rewriter.getContext());
+
+    // Linalg generic op for indexing the topk output idx tensor using
+    // the idx tensor returned by the linalg generic op for finding max.
+    // Only the idx tensor from the linalg generic op is sent as input.
+    auto extractedIdxLinalg = rewriter.create<linalg::GenericOp>(
+        loc, ArrayRef<Type>({filledTensorExtractedIdx.getType()}), findMaxIdx,
+        filledTensorExtractedIdx, extractedIdxMaps, extractedIdxIteratorTypes,
+        [&](OpBuilder &nestedBuilder, Location nestedLoc,
+            ValueRange blockArgs) {
+          // Get the current input idx.
+          Value index = rewriter.create<arith::IndexCastOp>(
+              loc, rewriter.getIndexType(), blockArgs[0]);
+
+          // Create idx to index the topk idx tensor.
+          // Index the dim dimension using the current input idx.
+          SmallVector<Value> indexTarget;
+          for (unsigned i = 0; i < dim; i++)
+            indexTarget.push_back(rewriter.create<linalg::IndexOp>(loc, i));
+          indexTarget.push_back(index);
+          for (unsigned i = dim; i < findMaxIdxType.getRank(); i++)
+            indexTarget.push_back(rewriter.create<linalg::IndexOp>(loc, i));
+
+          // Extract the element from the topk idx tensor.
+          Value extractedElement = rewriter.create<tensor::ExtractOp>(
+              loc, topkOpVal.back(), indexTarget);
+          rewriter.create<linalg::YieldOp>(loc, extractedElement);
+        });
+
+    auto extractedIdx = extractedIdxLinalg.getResult(0);
+    auto extractedIdxType = extractedIdx.getType().cast<RankedTensorType>();
+
+    // ======== END: Linalg generic for index extraction ========
+
+    // ======== BEGIN: Linalg generic for topk idx cast ========
+    // Casts from i32 to idx result type of the Kthvalue op.
+
+    // Create the initial tensor for the cast result.
+    Value filledTensorCastedIdx = createZeroInitTensor(
+        rewriter, loc, getTensorSizes(rewriter, loc, extractedIdx),
+        idxResultElementType);
+
+    SmallVector<utils::IteratorType> castedIdxIteratorTypes(
+        extractedIdxType.getRank(), utils::IteratorType::parallel);
+
+    SmallVector<AffineExpr> castedIdxMapExprs;
+    for (auto size : llvm::enumerate(
+             makeShapeTorchCompatible(extractedIdxType.getShape()))) {
+      castedIdxMapExprs.push_back(rewriter.getAffineDimExpr(size.index()));
+    }
+
+    auto castedIdxMaps = AffineMap::inferFromExprList(
+        {castedIdxMapExprs, castedIdxMapExprs}, rewriter.getContext());
+
+    // Linalg generic op for casting topk idx output tensor elements from i32 to
+    // result idx tensor element type.
+    auto castedIdxLinalg = rewriter.create<linalg::GenericOp>(
```

Comment:
Technically you can just simplify this by not bothering with `linalg::GenericOp`. If I remember correctly you can just do `arith::IndexCastOp` on tensors.

Reply:
Hey @rsuderman, thank for the comments!

I tried using `arith::IndexCastOp` directly on tensors, but I get the following error:

```
error: failed to legalize operation 'arith.index_cast'
note: see current operation: %14 = "arith.index_cast"(%13) : (tensor<3xi32>) -> tensor<3xindex>
error: Module does not conform to the linalg-on-tensors backend contract. See dialect conversion legality information above.
```

Basically, casting from a tensor of i32 element type to a tensor with the same shape and index element type fails. The docs state that the operation can be used for conversion between scalar or vector integers/indexes, but the operation is marked as Tensorizable (as all arith operations are), so it should pass for tensors too. Do you have an idea what could have produced this error?

Reply:
@rsuderman The problem here is that operations from the `arith` dialect will not be legalized by the `torch-verify-linalg-on-tensors-backend-contract` pass. The implementation of this pass adds `arith` as a legal dialect only for scalar ops:

```
auto isLegalScalarOp = [&](Operation *op) {
      // We recognize basic scalar ops by them having the trait "Elementwise",
      // even though we don't expect them to operate on tensors.
      return scalarConverter.isLegal(op) &&
             op->hasTrait<OpTrait::Elementwise>();
};

...

target.addDynamicallyLegalDialect<arith::ArithDialect>(isLegalScalarOp);
```

When I add the following to the pass:
```
target.addDynamicallyLegalDialect<arith::ArithDialect>(opHasLegalTypes);
```
the error dissapears. `opHasLegalTypes` is used in this pass for every dialect that works on tensors. So I guess the only way to not use linalg.generic is to modify the pass, but I don't know what other side-effects this can have, so I don't think it's the best idea for me to do this.

Is it alright if I stick with linalg.generic here?